web/src/main/resources/static/js/tree.js

```diff
@@ -39,15 +35,79 @@ var tree = (function () {
         this.acronym = '';
         this.mainType = '';
         this.color = '';
-        this.nci = '';
-        this.umls = '';
+        this.nci = [];
+        this.umls = [];
+        this.history = ''; // comma delimited string
+    }
+
+    function getOncotreeCodeKeysSortedByName(oncotreeNodeDict) {
+        return Object.keys(oncotreeNodeDict).sort(function(a,b) {
+                var nameA = oncotreeNodeDict[a].name;
+                var nameB = oncotreeNodeDict[b].name;
+                if (nameA < nameB) {
+                    return -1;
+                }
+                if (nameA > nameB) {
+                    return 1;
+                }
+                return 0;
+            });
+    }
+
+    function process_children(parentNode, childData) {
+        // childData is always for a new unique node
+        var childNode = new UniqueTreeNodeDatum();
+        childNode.name = childData.name + " (" + childData.code + ")";
+        childNode.acronym = childData.code;
+
+        if (childData.hasOwnProperty('mainType') && childData.mainType != null && childData.mainType.hasOwnProperty('name') && childData.mainType.name !== '') {
+            childNode.mainType = childData.mainType.name;
+        } else {
+            childNode.mainType = 'Not Available';
+        }
+
+        if (childData.hasOwnProperty('color')) {
+            childNode.color = childData.color;
+        }
+        if (childData.hasOwnProperty('NCI')) {
+            childNode.nci = childData.NCI;
+        }
+
+        if (childData.hasOwnProperty('UMLS')) {
+            childNode.umls = childData.UMLS;
+        }
+
+        if (childData.hasOwnProperty('history')) {
+            childNode.history = childData.history.join();
+        }
+
+        // save code and name to check for duplicate codes later
+        if (!oncotreeCodesToNames.hasOwnProperty(childNode.acronym)) {
+            oncotreeCodesToNames[childNode.acronym] = [];
+        }
+        oncotreeCodesToNames[childNode.acronym].push(childNode.name);
+
+        // add new node to children list of parentNode
+        parentNode.children.push(childNode);
+
+        // now process this node's children
+        if (childData.hasOwnProperty('children')) {
+            var codesSortedByName =  getOncotreeCodeKeysSortedByName(childData.children);
+            if (codesSortedByName.length > 0) {
+                childNode.children = [];
+                codesSortedByName.forEach(function (code) {
+                    // now childNode's children are added
+                    process_children(childNode, childData.children[code]);
```

Comment:
recursion: makes sense with our representation. If we ever create a topbraid relationship which creates a cycle in the tree we will be in trouble - but presumably it would then cease to be a "tree" and so should not happen.